base_invader/http.py

```diff
@@ -0,0 +1,126 @@
+
+    def get_headers(environ=None):
+        """Get a list of headers."""
+        _headers = [('Content-Type', 'application/json')]
+        for key, value in get_original_headers(environ=environ):
+            if key != 'Content-Type':
+                _headers.append(key, value)
+        return _headers
+
+    exception.get_body = get_body
+    exception.get_headers = get_headers
+    if request:
+        httprequest = request.httprequest
+        headers = dict(httprequest.headers)
+        headers.pop('Api-Key', None)
+        message = (
+            'RESTFULL call to url %s with method %s and params %s '
+            'raise the following error %s')
+        args = (httprequest.url, httprequest.method, request.params, exception)
+        extra = {
+            'application': request.backend.name,
+            'url': httprequest.url,
+            'method': httprequest.method,
+            'params': request.params,
+            'headers': headers,
+            'status': exception.code,
+            'exception_body': exception.get_body(),
+            'traceback': ''.join(traceback.format_exception(*sys.exc_info())),
+            }
+        _logger.error(message, *args, extra=extra)
+    return exception
+
+
+class HttpRestRequest(HttpRequest):
+    """Http request that always return json, usefull for rest api"""
+
+    def __init__(self, *args):
+        super(HttpRestRequest, self).__init__(*args)
+        if self.httprequest.headers.get('Content-Type') == 'application/json':
+            self.params = json.loads(self.httprequest.stream.read())
+
+    def _handle_exception(self, exception):
+        """Called within an except block to allow converting exceptions
+           to abitrary responses. Anything returned (except None) will
+           be used as response."""
+        _logger.exception('Shopinvader Handle exception %s', exception)
+        try:
+            return super(HttpRestRequest, self)._handle_exception(exception)
+        except (UserError, ValidationError), e:
+            return wrapJsonException(
+                BadRequest(e.message or e.value or e.name))
+        except MissingError, e:
+            return wrapJsonException(NotFound(e.value))
+        except AccessError, e:
+            return wrapJsonException(Forbidden(e.value))
+        except HTTPException, e:
+            return wrapJsonException(e)
+        except:  # flake8: noqa: E722
+            return wrapJsonException(InternalServerError())
+
+    def make_json_response(self, data, headers=None, cookies=None):
+        data = JSONEncoder().encode(data)
+        if headers is None:
+            headers = {}
+        headers['Content-Type'] = 'application/json'
+        return self.make_response(data, headers=headers, cookies=cookies)
+
+
+ori_get_request = Root.get_request
+
+
+def get_request(self, httprequest):
+    if httprequest.path.startswith(ROOT_PATH):
+        return HttpRestRequest(httprequest)
+    return ori_get_request(self, httprequest)
```

Comment:
Instead of patching HttpRequest, could we do this in `_process_method`? It would make it easier to define alternative controllers (eg with different auth methods) while benefiting for this json hanlding.

Reply:
I'll check again but unfortunately this place into Odoo is not designed to be extended.